### PR TITLE
feat(decompile): support rendering ibis expression as python code

### PR DESCRIPTION
**Pull request by kszucs**
depends on #4493 

Closes https://github.com/ibis-project/ibis/issues/4335.

### CODE REVIEW

File: ibis/common/graph.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        queue.extend(children)
+
+    return graph
+
+
+def dfs(node: Traversable, types=Traversable):
+    assert issubclass(types, Traversable)
+    assert isinstance(node, types)
+
+    stack = deque([node])
```
Comment (cpcloud) [inline ibis/common/graph.py:45]: No reason to use a `deque` here.

```suggestion
    stack = [node]
```

File: ibis/common/graph.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+"""
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections import deque
+from typing import Any, Callable, Iterable, Iterator
+
+import toolz
+
+# TODO(kszucs): add recursive implementation for bfs and dfs
```
Comment (cpcloud) [inline ibis/common/graph.py:12]: Do we need to?

File: ibis/common/graph.py (hunk 1 above)
Comment (kszucs) [reply ibis/common/graph.py:46]: deque is supposed to be slightly more performant in this scenario due to frequent subsequend adding and popping

File: ibis/common/validators.py [hunk 3]
```diff
@@ -8,18 +8,26 @@
...
+from ibis.util import DotDict, flatten_iterable, is_function, is_iterable
 
 EMPTY = inspect.Parameter.empty  # marker for missing argument
+VAR_POSITIONAL = inspect.Parameter.VAR_POSITIONAL
+POSITIONAL_OR_KEYWORD = inspect.Parameter.POSITIONAL_OR_KEYWORD
 
 
-class Validator(Callable):
+# TODO(kszucs): rename Validator to Annotation
+class Annotation(Callable):
```
Comment (kszucs) [inline ibis/common/validators.py:19]: I'd prefer to rename this module to `annotations.py`

File: ibis/common/validators.py [hunk 4]
```diff
@@ -48,12 +59,41 @@ def __call__(self, arg, **kwargs):
...
+    Marker class for validating variadic arguments.
+    """
+
+    __slots__ = ('validator',)
+
+    def __init__(self, validator):
+        self.validator = validator
+
+    def __repr__(self):
+        return f"Variadic({self.validator!r})"
```
Comment (cpcloud) [inline ibis/common/validators.py:81]: Might as well use `self.__class__` here.

```suggestion
        return f"{self.__class__}({self.validator!r})"
```

File: ibis/common/validators.py (hunk 3 above)
Comment (cpcloud) [reply ibis/common/validators.py:19]: Sounds good to me!

File: ibis/common/tests/test_graph.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    assert g == {D: [], E: [], B: [D, E], C: [], A: [B, C]}
+
+    i = invert(g)
+    assert i == {D: [B], E: [B], B: [A], C: [A], A: []}
+
+    j = invert(i)
+    assert j == g
+
+
+def test_toposort():
```
Comment (cpcloud) [inline ibis/common/tests/test_graph.py:52]: What happened to the test cases for edge cases like an empty graph?

File: ibis/common/grounds.py [hunk 6]
```diff
@@ -244,3 +260,14 @@ def args(self):
     @property
     def argnames(self):
         return self.__argnames__
+
+    def map(self, fn):
```
Comment (cpcloud) [inline ibis/common/grounds.py:264]: Looks like this method could use a test or two.

File: ibis/common/grounds.py [hunk 7]
```diff
@@ -219,15 +232,18 @@ def __cached_equals__(self, other):
         return result
 
 
-class Concrete(Immutable, Comparable, Annotable):
+# TODO(kszucs): consider to move it somewhere else, like ibis/expr/core.py
+class Concrete(Immutable, Comparable, Traversable, Annotable):
```
Comment (cpcloud) [inline ibis/common/grounds.py:236]: It doesn't seem desirable to have every ibis object be traversible. I would expect `Traversable` to exclude data types and schemas and effectively be a filter for `ops.Node`. Why would data types and schemas ever need to be traversable in the same way that `Node`s are?

File: ibis/common/grounds.py [hunk 8]
```diff
@@ -201,7 +214,7 @@ def __cached_equals__(self, other):
             return True
 
         # type comparison should be cheap
-        if type(self) != type(other):
+        if type(self) is not type(other):
```
Comment (cpcloud) [inline ibis/common/grounds.py:217]: Is this change meaningful in some way? When is `!=` not equivalent to `is not` in this context? Dynamically generated types?

File: ibis/backends/decompile/compiler.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+def literal(op, value, dtype):
+    return repr(value)
+
+
+@translate.register(ops.NodeList)
+def node_list(op, *values):
+    return ", ".join(values)
+
+
+@translate.register(ops.Equals)
```
Comment (cpcloud) [inline ibis/backends/decompile/compiler.py:87]: Instead of defining each of the non-syntactic APIs, e.g., `StringColumn.split()` can we have a generic `Value` implementation that looks at a new class property and if that's `None` (the default) snake cases the operation class name to be the method call?

Otherwise it's a super annoying game of whack-a-mole for every operation.

File: ibis/backends/decompile/compiler.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+
+@translate.register(dt.DataType)
+def datatype(dtype, *args, **kwargs):
+    return str(dtype)
+
+
+@translate.register(sch.Schema)
+def schema(schema, *, names, types):
+    return repr(list(zip(names, map(str, types))))
```
Comment (cpcloud) [inline ibis/backends/decompile/compiler.py:25]: Can we use the `dict` literal representation here? The list of tuples repr is super noisy vs the dict.

File: ibis/backends/decompile/compiler.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+
+@translate.register(ops.TableColumn)
+def table_column(op, table, name):
+    return f"{table}.{name}"
+
+
+@translate.register(ops.SortKey)
+def sort_key(op, expr, ascending):
+    if ascending:
+        return f"ibis.asc({expr})"
```
Comment (cpcloud) [inline ibis/backends/decompile/compiler.py:61]:
```suggestion
        return f"{expr}.asc()"
```

File: ibis/backends/decompile/compiler.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    else:
+        return f"ibis.desc({expr})"
```
Comment (cpcloud) [inline ibis/backends/decompile/compiler.py:63]:
```suggestion
        return f"{expr}.desc()"
```

File: ibis/backends/decompile/compiler.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+def node_list(op, *values):
+    return ", ".join(values)
+
+
+@translate.register(ops.Equals)
+def equals(op, left, right):
+    return f"{left} == {right}"
+
+
+@translate.register(ops.Divide)
```
Comment (cpcloud) [inline ibis/backends/decompile/compiler.py:92]: Eventually the arithmetic operations will have to be parenthesized. I know you haven't implemented `Add` or `Subtract` yet, but you won't be able to just render the operands recursively.

Consider this case:

```
In [11]: (ibis.literal(2) - ibis.literal(3)) + 4
Out[11]: Add(Subtract(2, 3), 4): 2 - 3 + 4

In [12]: ibis.literal(2) - (ibis.literal(3) + 4)
Out[12]: Subtract(2, Add(3, 4)): 2 - 3 + 4
```

We're technically rendering this incorrectly in the repr right now, and you'll have to deal with the parenthesization problem here too.

**Review by cpcloud — CHANGES_REQUESTED** [review]
Looking good! A few more things to tackle before we can merge.

Can we move this out of the `backends` directory? Backends do things like create tables and execute expressions and this does neither of those things.

How about moving it to the top level, and adding either a top level `ibis.decompile()` method or `expr.decompile()` method?

File: ibis/common/validators.py (hunk 3 above)
Comment (kszucs) [reply ibis/common/validators.py:19]: Renamed in #4552

File: ibis/backends/decompile/compiler.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+class Varnames:
+    def __init__(self):
+        self.t = itertools.count()
+        self.v = itertools.count()
+
+    def generate(self, node):
+        if isinstance(node, ops.TableNode):
+            return f"t{next(self.t)}"
+        elif isinstance(node, sch.Schema):
+            return f"s{next(self.s)}"
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/decompile/compiler.py:106]: Forgot to initialize it, fixing.

File: ibis/backends/decompile/compiler.py [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+
+
+@functools.singledispatch
+def translate(op, *args, **kwargs):
+    raise NotImplementedError(op)
+
+
+@translate.register(dt.DataType)
+def datatype(dtype, *args, **kwargs):
+    return str(dtype)
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/decompile/compiler.py:20]: IIRC `str()` should place `!` for non-nullable datatypes

File: ibis/backends/decompile/compiler.py [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+
+
+@translate.register(sch.Schema)
+def schema(schema, *, names, types):
+    return repr(list(zip(names, map(str, types))))
+
+
+@translate.register(ops.UnboundTable)
+def table(op, schema, name):
+    return f"ibis.table(name={name!r}, schema={schema})"
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/decompile/compiler.py:30]: Done.

File: ibis/common/grounds.py [hunk 17]
```diff
@@ -150,6 +161,9 @@ def __eq__(self, other):
     def __getstate__(self):
         return {name: getattr(self, name) for name in self.__argnames__}
 
+    def __dismantle__(self, prefer_keywords=False):
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/common/grounds.py]: Removed it

File: ibis/common/validators.py (hunk 4 above)
Comment (kszucs) [reply ibis/common/validators.py:81]: Not relevant of this PR's scope anymore.

File: ibis/common/grounds.py (hunk 6 above)
Comment (kszucs) [reply ibis/common/grounds.py:268]: Moved the comment to the other PR

File: ibis/common/grounds.py (hunk 8 above)
Comment (kszucs) [reply ibis/common/grounds.py:215]: It's just more expressive syntactically.

File: ibis/backends/decompile/compiler.py (hunk 11 above)
Comment (kszucs) [reply ibis/backends/decompile/compiler.py:61]: Updated.

File: ibis/backends/decompile/compiler.py (hunk 12 above)
Comment (kszucs) [reply ibis/backends/decompile/compiler.py:63]: Updated.

File: ibis/backends/decompile/compiler.py [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+
+
+@translate.register(ops.Alias)
+def alias(op, arg, name):
+    return f"{arg}.name({name!r})"
+
+
+@translate.register(ops.Literal)
+def literal(op, value, dtype):
+    return f"ibis.literal({value!r}, \"{dtype}\")"
```
Comment (mik-laj) [inline ibis/backends/decompile/compiler.py:85]:
```suggestion
    return f"ibis.literal({value!r}, {dtype!r})"
```

File: ibis/backends/decompile/compiler.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+        elif n_dependents == 1:
+            return result
+        else:
+            out.write(f"expr = {result}\n")
+            return result
+
+    node.map(fn)
+    # graph.map(fn, node)
+
+    rendered = f"import ibis\n{out.getvalue()}"
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/decompile/compiler.py:170]: Added an option.

File: ibis/backends/decompile/compiler.py (hunk 10 above)
Comment (kszucs) [reply ibis/backends/decompile/compiler.py]: Updated.

File: ibis/backends/decompile/compiler.py (hunk 18 above)
Comment (kszucs) [reply ibis/backends/decompile/compiler.py:85]: Updated.

File: ibis/backends/decompile/compiler.py (hunk 9 above)
Comment (kszucs) [reply ibis/backends/decompile/compiler.py]: @cpcloud I may misunderstand, do you mean we should introduce a new class property?

**Comment by kszucs** [comment]
I'd recommend marking this as an experimental feature since a couple of expressions don't roundtrip properly, most notably the reductions rewritten to aggregations. 

On the other hand I managed to reach a rather high coverage using the already removed sqlgolden fixture to exercise `decompile()` on multiple expressions (see the temporarily committed, generated python files). @cpcloud could you recommend a method to properly test the decompiler without manually maintaining a bunch of tests? 

I'm not planning to invest much more time to this PR other than finalizing it. 

cc @saulpw @jcrist

File: ibis/backends/decompile/compiler.py (hunk 13 above)
Comment (kszucs) [reply ibis/backends/decompile/compiler.py]: I'm handling parenthesization from parent infix operators now, seems to work fine.

**Comment by cpcloud** [comment]
@kszucs Can you use `pytest-snapshot` for the tests here? It's hard to see if you are because the large number of files changed makes the browser slow 😬

**Comment by kszucs** [comment]
@cpcloud updated, could you take another look?

File: ibis/expr/decompile.py [hunk 20]
```diff
@@ -0,0 +1,409 @@
...
+
+    node.map(fn)
+
+    result = out.getvalue()
+    if render_import:
+        result = f"import ibis\n\n\n{result}"
+    if format:
+        import black
+
+        result = black.format_str(result, mode=black.FileMode())
```
Comment (cpcloud) [inline ibis/expr/decompile.py:407]: Will this pickup configuration using black's default `pyproject.toml` search procedure?

File: ibis/tests/sql/conftest.py [hunk 21]
```diff
@@ -4,6 +4,60 @@
...
+# def sqlgolden(func):  # decorator
+#     @functools.wraps(func)
+#     def _decorated(*args, **kwargs):
+#         sql_golden_check(func.__name__, func(*args, **kwargs))
+#         ibis_golden_check(func.__name__, func(*args, **kwargs))
+
+#     return _decorated
+
+
+# def sqlgoldens(func):  # decorator
```
Comment (cpcloud) [inline ibis/tests/sql/conftest.py:53]: I think all this commented out code can be deleted.

File: ibis/tests/sql/snapshots/test_compiler/test_column_expr_default_name/decompiled.py [hunk 22]
```diff
@@ -0,0 +1,4 @@
+import ibis
+
+
+result = ibis.table(name="int_col_table", schema={"int_col": "int32"}).int_col + 4
```
Comment (cpcloud) [inline ibis/tests/sql/snapshots/test_compiler/test_column_expr_default_name/decompiled.py:4]: Should this be generating an intermediate table variable?

File: ibis/tests/sql/snapshots/test_select_sql/test_case_in_projection/decompiled.py [hunk 23]
```diff
@@ -0,0 +1,40 @@
+import ibis
+
+
+l = ibis.literal("foo")
```
Comment (cpcloud) [inline ibis/tests/sql/snapshots/test_select_sql/test_case_in_projection/decompiled.py:4]: Not a huge fan of variable names that start with `l`, but it's generated code so 🤷🏻

**Review by cpcloud — APPROVED** [review]
Brilliant start!

🚢 it.

File: ibis/backends/decompile/compiler.py (hunk 9 above)
Comment (cpcloud) [reply ibis/backends/decompile/compiler.py]: This is resolved!

File: ibis/tests/sql/snapshots/test_compiler/test_column_expr_default_name/decompiled.py (hunk 22 above)
Comment (kszucs) [reply ibis/tests/sql/snapshots/test_compiler/test_column_expr_default_name/decompiled.py:4]: Currently when an expression gets used once it doesn't generate an intermediate variable, though we can add exceptions based on the types.

File: ibis/tests/sql/conftest.py (hunk 21 above)
Comment (kszucs) [reply ibis/tests/sql/conftest.py:53]: Removed.

File: ibis/tests/sql/snapshots/test_compiler/test_column_expr_default_name/decompiled.py (hunk 22 above)
Comment (cpcloud) [reply ibis/tests/sql/snapshots/test_compiler/test_column_expr_default_name/decompiled.py:4]: Ah, ok. I think this is fine as is.

Comment (kszucs) [reply ibis/tests/sql/snapshots/test_compiler/test_column_expr_default_name/decompiled.py:4]: Updated.

File: ibis/expr/decompile.py (hunk 20 above)
Comment (kszucs) [reply ibis/expr/decompile.py:407]: Great question, I have not checked it.

File: ibis/tests/sql/snapshots/test_select_sql/test_case_in_projection/decompiled.py (hunk 23 above)
Comment (kszucs) [reply ibis/tests/sql/snapshots/test_select_sql/test_case_in_projection/decompiled.py:4]: Renamed it to `lit` for now

**Comment by kszucs** [comment]
@cpcloud addressed the review comments and also did some refactoring

**Comment by cpcloud** [comment]
Sweet, merging!
